libcodechecker/analyze/tu_collector.py

```diff
@@ -0,0 +1,257 @@
+                         toolchain then the toolchain compiler's dependencies
+                         are also collected in case this parameter is True.
+    """
+
+    logging.debug("Generating dependent headers via compiler...")
+
+    if isinstance(command, str) or isinstance(command, unicode):
+        command = shlex.split(command)
+
+    dependencies = set()
+    error = ''
+
+    try:
+        dependencies |= set(__gather_dependencies(command, build_dir))
+    except Exception as ex:
+        logging.debug("Couldn't create dependencies:")
+        logging.debug(str(ex))
+        error += str(ex)
+
+    toolchain_compiler = __get_toolchain_compiler(command)
+
+    if collect_toolchain and toolchain_compiler:
+        logging.debug("Generating gcc-toolchain headers via toolchain "
+                      "compiler...")
+        try:
+            # Change the original compiler to the compiler from the toolchain.
+            command[0] = toolchain_compiler
+            dependencies |= set(__gather_dependencies(command, build_dir))
+        except Exception as ex:
+            logging.debug("Couldn't create dependencies:")
+            logging.debug(str(ex))
+            error += str(ex)
+
+    return dependencies, error
+
+
+def add_sources_to_zip(zip_file, files):
+    """
+    This function adds source files to the ZIP file if those are not present
+    yet. The files will be placed to the "sources-root" directory under the ZIP
+    file.
+
+    zip_file -- A ZIP file name as a string.
+    files -- A file path or an iterable of file paths.
+    """
+    if isinstance(files, str):
+        files = [files]
+
+    with zipfile.ZipFile(zip_file, 'a') as archive:
+        for f in files:
+            archive_path = os.path.join('sources-root', f.lstrip('/'))
+
+            try:
+                archive.getinfo(archive_path)
+            except KeyError:
+                archive.write(f, archive_path, zipfile.ZIP_DEFLATED)
+            else:
+                logging.debug("'{}' is already in the ZIP file, won't add it "
+                              "again!".format(f))
+
+
+def zip_tu_files(zip_file, compilation_database, write_mode='w'):
+    """
+    Collects all files to a zip file which are required for the compilation of
+    the translation units described by the given compilation database.
+    If there are some files which couldn't be collected then a "no-sources"
+    file will contain the error message of its reason.
+    The function returns the set of files.
+
+    zip_file -- A file name or a file object.
+    compilation_database -- Either a path of the compilation database JSON file
+                            or a list of the parsed JSON.
+    write_mode -- The file opening mode of the zip_file. In case of 'a' the new
+                  files are appended to the existing zip file, in case of 'w'
+                  the files are added to a clean zip file.
+    """
+    if isinstance(compilation_database, str) or \
```

Comment:
We use this type checking multiple places:
```py
isinstance(data, str) or isinstance(data, unicode)
```
Maybe we could create a helper function for this.